File: 2d_matrix.java
// Time Complexity : O(log(m*n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach in three sentences only
// Find appropriate row and colum for given index of element.
// perform binary search by getting element matrix[row][column]
class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        if(matrix.length == 0 || matrix ==null){
            return false;
        }
        int m = matrix.length; // rows
        int n = matrix[0].length; //columns
        int low = 0;
        int high = m*n -1;

        while(low <= high){
            int mid = low + high /2;
            int row = mid / n;
            int col = mid % n;

            if(matrix[row][col] == target){
                return true;
            }
            else if (target > matrix[row][col]){
                low = mid + 1;
            }
            else{
                high = mid - 1;
            }
        }
        return false;
    }
}

File: 2dMatrix.py
# time: O(log mn)
# space: O(1)

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        top, bottom = 0, len(matrix)-1
        l,r = 0, len(matrix[0])

        while top <= bottom:
            mid = (top+bottom)//2
            if target>=matrix[mid][0] and target<=matrix[mid][-1]:
                break
            elif target<matrix[mid][0]:
                bottom-=1
            else:
                top+=1

        while l<r:
            m = (l+r)//2
            if target<matrix[mid][m]:
                r-=1
            elif target>matrix[mid][m]:
                l+=1
            elif target==matrix[mid][m]:
                return True

        return False
